addons/storysource/src/StoryPanel.tsx

```diff
@@ -0,0 +1,204 @@
+import React from 'react';
+import { API } from '@storybook/api';
+import { styled } from '@storybook/theming';
+import { Link } from '@storybook/router';
+import {
+  SyntaxHighlighter,
+  SyntaxHighlighterProps,
+  SyntaxHighlighterRendererProps,
+} from '@storybook/components';
+
+import createElement from 'react-syntax-highlighter/dist/esm/create-element';
+
+const StyledStoryLink = styled(Link)<{ to: string; key: string }>(({ theme }) => ({
+  display: 'block',
+  textDecoration: 'none',
+  borderRadius: theme.appBorderRadius,
+  color: 'inherit',
+
+  '&:hover': {
+    background: theme.background.hoverable,
+  },
+}));
+
+const SelectedStoryHighlight = styled.div(({ theme }) => ({
+  background: theme.background.hoverable,
+  borderRadius: theme.appBorderRadius,
+}));
+
+const StyledSyntaxHighlighter = styled(SyntaxHighlighter)<SyntaxHighlighterProps>(({ theme }) => ({
+  fontSize: theme.typography.size.s2 - 1,
+}));
+
+interface SourceLoc {
+  line: number;
+  col: number;
+}
+
+interface SourceBlock {
+  startLoc: SourceLoc;
+  endLoc: SourceLoc;
+}
+
+const areLocationsEqual = (a: SourceBlock, b: SourceBlock): boolean =>
+  a.startLoc.line === b.startLoc.line &&
+  a.startLoc.col === b.startLoc.col &&
+  a.endLoc.line === b.endLoc.line &&
+  a.endLoc.col === b.endLoc.col;
+
+interface LocationsMap {
+  [key: string]: SourceBlock;
+}
+
+interface StoryPanelProps {
+  api: API;
+}
+
+interface SourceParams {
+  source: string;
+  locationsMap: LocationsMap;
+}
+export interface StoryData {
+  id: string;
+  kind?: string;
+  parameters?: {
+    storySource?: SourceParams;
+    mdxSource?: string;
+  };
+}
+export const StoryPanel: React.FC<StoryPanelProps> = ({ api }) => {
+  const [state, setState] = React.useState<SourceParams & { currentLocation?: SourceBlock }>({
+    source: 'loading source...',
```

Comment:
is there any way to detect when the source-loader failed and show an error message?

Reply:
maybe if no source code is found at all in the storySourc parameter?